[Core] Add healthstone/health pots to spellbook and death recap.

Gebuz opened the pull request:
Also fixed an issue with death recap not working with multi id spells in the spellbook (such as healthstone/pot).
![image](https://user-images.githubusercontent.com/6773230/40891306-806fa43c-6783-11e8-8ad4-c3de53647c60.png)
![image](https://user-images.githubusercontent.com/6773230/40891310-8ac92c96-6783-11e8-92d2-d36b268714df.png)

MartijnHols commented on src/Main/DeathRecap.js:
```diff
@@ -201,15 +201,15 @@ class DeathRecap extends React.PureComponent {
                         <td style={{ width: '20%' }}>
                           {event.buffsUp && event.buffsUp.sort(sortByTimelineIndex).map(e =>
-                            <SpellIcon style={{ border: '1px solid rgba(0, 0, 0, 0)' }} id={e.spell ? e.spell.id : e} />
+                            <SpellIcon style={{ border: '1px solid rgba(0, 0, 0, 0)' }} id={e.spell instanceof Array ? e.spell[0].id : e.spell.id} />
```
Please use primarySpell instead https://github.com/WoWAnalyzer/WoWAnalyzer/blob/master/src/Parser/Core/Modules/Ability.js#L108

Gebuz replied:
Not sure why, but this doesn't work :( will try to figure out why tomorrow.

MartijnHols commented on src/Main/DeathRecapTracker.js:
```diff
@@ -44,14 +44,16 @@ class DeathRecapTracker extends Analyzer {
     const cooldownsOnly = this.cooldowns.filter(e => e.cooldown);
-    extendedEvent.defensiveCooldowns = cooldownsOnly.map(e => ({...e, cooldownReady: this.spellUsable.isAvailable(e.spell.id)}));
+    extendedEvent.defensiveCooldowns = cooldownsOnly.map(e => {
+      return ({...e, cooldownReady: this.spellUsable.isAvailable(e.spell instanceof Array ? e.spell[0].id : e.spell.id)});
```
Return can be omitted when there’s only a single statement.

MartijnHols commented on src/Parser/Core/Modules/Items/Healthstone.js:
```diff
@@ -0,0 +1,42 @@
+    if (!this.spellUsable.isOnCooldown(HEALTHSTONE_SPELLS[0].id)){
+      return;
+    }
+    this.spellUsable.reduceCooldown(HEALTHSTONE_SPELLS[0].id, COOLDOWN_MS - OUT_OF_COMBAT_COOLDOWN_MS);
```
Death doesn’t drop combat (not since wotlk)

Gebuz replied:
Deaths still starts Healthstone/pot cooldowns however.

MartijnHols commented on src/Parser/Core/Modules/Items/Healthstone.js:
```diff
@@ -0,0 +1,42 @@
+import Abilities from 'Parser/Core/Modules/Abilities';
+import SpellUsable from 'Parser/Core/Modules/SpellUsable';
+
+const HEALTHSTONE_SPELLS = [
```
Could this or prepotion somehow work together? I don’t want to have to maintain two identical lists.

Gebuz replied:
HS/healthpots don't share cooldowns with DPS/mana pots. Shouldn't be any overlaps in the lists.

MartijnHols commented on src/Parser/Core/CombatLogParser.js:
```diff
@@ -269,6 +271,9 @@ class CombatLogParser {
     umbralMoonglaives: UmbralMoonglaives,
+
+    //added last so that it includes items that are added to the spellbook on initialization.
+    deathRecapTracker: DeathRecapTracker,
```
This doesn’t work like this. Import order is determined by dependencies, anything with a single dependency not loaded will be loaded on a later iteration.

Gebuz replied:
The DeathRecapTracker sets the list on IDs to track in it's 'on_initialize'. There's probably a better way to handle it but I didn't want to mess around too much with it, and it seemed to consistently not include Healthstone if the Healthstone module was after it in the dependency list, and likewise always include it when HS module was before it.

Gebuz replied:
Okay I'm dumb, just added healthstone as a dependency to the death recap tracker <.<

MartijnHols commented on src/Parser/Monk/Brewmaster/Modules/Features/Abilities.js:
```diff
@@ -69,13 +69,13 @@ class Abilities extends CoreAbilities {
         spell: [SPELLS.IRONSKIN_BREW, SPELLS.PURIFYING_BREW],
         buffSpellId: SPELLS.IRONSKIN_BREW_BUFF.id,
-        category: Abilities.SPELL_CATEGORIES.COOLDOWNS,
+        category: Abilities.SPELL_CATEGORIES.DEFENSIVE,
```
These seems like a miscategorization, I recommend using isDefensive instead.

Gebuz replied:
Why? The ability is purely defensive.

MartijnHols commented on src/Parser/Core/Modules/Items/Healthstone.js:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  on_toPlayer_death(event) {
+    if (!this.spellUsable.isOnCooldown(HEALTHSTONE_SPELLS[0].id)){
```
You can’t expect the first index to always be the same thing.

Gebuz commented:
Turns out the reason .primarySpell didn't work was because sometimes it was only the `spell` that was added instead of the full `abilities` entrance. I changed it to only add the ID (and for cooldowns, also `cooldownReady`) to the buffs, debuffs and defensives array instead of the full abilities/spell/debuff event. Still missing some other stuff.

Gebuz commented:
Not correctly starts the cooldown at 60s on death:
![image](https://user-images.githubusercontent.com/6773230/40933152-f39bc212-6830-11e8-8da4-8f71087b3d74.png)
Also calculates max casts correctly:
![image](https://user-images.githubusercontent.com/6773230/40933170-02f4279a-6831-11e8-8eb2-3f172c31362d.png)

MartijnHols commented on src/Main/DeathRecapTracker.js:
```diff
@@ -44,15 +46,17 @@ class DeathRecapTracker extends Analyzer {
+    extendedEvent.defensiveCooldowns = cooldownsOnly.map(e => ({id: e.primarySpell.id, cooldownReady: this.spellUsable.isAvailable(e.primarySpell.id)}));
     if (event.hitPoints > 0) {
-      this.lastBuffs = this.buffs.filter(e => this.combatants.selected.hasBuff(e.buffSpellId) || this.combatants.selected.hasBuff(e.spell.id));
+      this.lastBuffs = this.buffs.filter(e => this.combatants.selected.hasBuff(e.buffSpellId) || this.combatants.selected.hasBuff(e.spell instanceof Array ? e.spell[0].id : e.spell.id))
```
Inconsistent spell id usage

MartijnHols commented on src/Parser/Core/Modules/Items/Healthstone.js:
```diff
@@ -0,0 +1,55 @@
+];
+
+const COOLDOWN_MS = 3600000; // one hour
+const OUT_OF_COMBAT_COOLDOWN_MS = 60000; // one minute
```
In that case please leave a comment explaining how this works ingame (or preferably add a test for this requirement).

Also if it’s triggered by dying, how about  naming it so that it refers to that instead of leaving combat?

MartijnHols commented on src/Parser/Core/Modules/Items/Healthstone.js:
```diff
@@ -0,0 +1,55 @@
+      return;
+    }
+    this.spellUsable.reduceCooldown(SPELLS.HEALTHSTONE.id, cooldownRemaining - OUT_OF_COMBAT_COOLDOWN_MS);
+    if(event.timestamp + OUT_OF_COMBAT_COOLDOWN_MS < this.owner.fight.end_time){
```
How about putting the future time in a var? With good naming this might become easier to understand. A comment explaining why might also clarify things.

Gebuz commented:
Changed `this.buffs` to be consistent in what it contains to remove the last `e.spell instanceof Array ? e.spell[0].id : e.spell.id`. It now only contains an object with `id`. I decided to leave it as an object so that more info like stacks etc. can be added in the future.

Also renamed the constants in Healthstone and added more comments.

Gebuz commented:
Aaaand it should now also calculate maxCasts correctly when HS was not used before a death. Also changed the efficiency thresholds.
